java/generate-offline-map/src/main/java/com/esri/arcgisruntime/generateofflinemap/MainActivity.java

```diff
@@ -0,0 +1,246 @@
+import com.esri.arcgisruntime.tasks.offlinemap.GenerateOfflineMapResult;
+import com.esri.arcgisruntime.tasks.offlinemap.OfflineMapTask;
+
+public class MainActivity extends AppCompatActivity {
+
+  private static final String TAG = MainActivity.class.getSimpleName();
+
+  private MapView mMapView;
+  private Button mTakeMapOfflineButton;
+  private GraphicsOverlay mGraphicsOverlay;
+  private Graphic mDownloadArea;
+
+  @Override
+  protected void onCreate(Bundle savedInstanceState) {
+    super.onCreate(savedInstanceState);
+    setContentView(R.layout.activity_main);
+
+    // access MapView from layout
+    mMapView = findViewById(R.id.mapView);
+
+    // access button to take the map offline and disable it until map is loaded
+    mTakeMapOfflineButton = findViewById(R.id.takeMapOfflineButton);
+    mTakeMapOfflineButton.setEnabled(false);
+
+    // request write permission
+    String[] reqPermission = { Manifest.permission.WRITE_EXTERNAL_STORAGE };
+    int requestCode = 2;
+    // for API level 23+ request permission at runtime
+    if (ContextCompat.checkSelfPermission(this, reqPermission[0]) == PackageManager.PERMISSION_GRANTED) {
+      generateOfflineMap();
+    } else {
+      // request permission
+      ActivityCompat.requestPermissions(this, reqPermission, requestCode);
+    }
+
+    // handle authentication with the portal
+    AuthenticationManager.setAuthenticationChallengeHandler(new DefaultAuthenticationChallengeHandler(this));
+
+    // create a portal item with the itemId of the web map
+    Portal portal = new Portal(getString(R.string.portal_url), true);
+    PortalItem portalItem = new PortalItem(portal, getString(R.string.item_id));
+
+    // create a map with the portal item
+    ArcGISMap map = new ArcGISMap(portalItem);
+    map.addDoneLoadingListener(() -> {
+      if (map.getLoadStatus() == LoadStatus.LOADED) {
+
+        // enable the map offline button only after the map is loaded
+        mTakeMapOfflineButton.setEnabled(true);
+
+        // limit the map scale to the largest layer scale
+        map.setMaxScale(map.getOperationalLayers().get(6).getMaxScale());
+        map.setMinScale(map.getOperationalLayers().get(6).getMinScale());
+      }
+    });
+
+    // set the map to the map view
+    mMapView.setMap(map);
+
+    // create a graphics overlay for the map view
+    mGraphicsOverlay = new GraphicsOverlay();
+    mMapView.getGraphicsOverlays().add(mGraphicsOverlay);
+
+    // create a graphic to show a box around the extent we want to download
+    mDownloadArea = new Graphic();
+    mGraphicsOverlay.getGraphics().add(mDownloadArea);
+    SimpleLineSymbol simpleLineSymbol = new SimpleLineSymbol(SimpleLineSymbol.Style.SOLID, Color.RED, 2);
+    mDownloadArea.setSymbol(simpleLineSymbol);
+
+    // update the download area box whenever the viewpoint changes
+    mMapView.addViewpointChangedListener(viewpointChangedEvent -> {
+      if (map.getLoadStatus() == LoadStatus.LOADED) {
```

Comment:
You could add an else-statement if the map fails to load

Reply:
Good idea!